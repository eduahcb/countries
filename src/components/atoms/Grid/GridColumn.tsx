import React, { ReactElement, ReactNode } from 'react'
import styled, { css } from 'styled-components'

const Root = styled.div<GridColumnProps>`
  width: 100%;
  grid-column: span ${props => props.xs ?? 1};

  ${
    props => props.sm && css`
      @media (min-width: 640px) {
        grid-column: span ${props.sm};
      }
    `
  }

  ${
    props => props.md && css`
      @media (min-width: 768px) {
        grid-column: span ${props.md};
      }
    `
  }

  ${
    props => props.lg && css`
      @media (min-width: 1024px) {
        grid-column: span ${props.lg};
      }
    `
  }

  ${
    props => props.xl && css`
      @media (min-width: 1440px) {
        grid-column: span ${props.xl};
      }
    `
  }
`

type GridColumnProps = {
  children: ReactNode
  xs?: number
  sm?: number
  md?: number
  lg?: number
  xl?: number
  style?: number
}

export const GridColumn = ({ children, xs, sm, md, lg, xl, style }: GridColumnProps): ReactElement => {
  return (
    <Root xs={xs} sm={sm} md={md} lg={lg} xl={xl} style={style}>{ children }</Root>
  )
}
