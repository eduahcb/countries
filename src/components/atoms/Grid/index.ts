export { Grid } from './Grid'
export { GridColumn } from './GridColumn'
