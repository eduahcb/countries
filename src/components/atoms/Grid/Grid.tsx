import React, { HTMLAttributes, ReactElement, ReactNode } from 'react'
import styled, { css } from 'styled-components'

import { GridColumn } from './GridColumn'

const Root = styled.div<GridProps>`
  width: 100%;
  display: grid;
  grid-template-columns: repeat(12, 1fr);
  
  ${
    props => props.gap && css`
      gap: ${props.gap}px;
    `
  }
`

type GridProps = HTMLAttributes<HTMLDivElement> & {
  children: ReactNode
  gap?: number
}

const Grid = ({ children, gap, ...rest }: GridProps): ReactElement => {
  return (
    <Root gap={gap} {...rest}>{ children }</Root>
  )
}

Grid.Column = GridColumn

export { Grid }
