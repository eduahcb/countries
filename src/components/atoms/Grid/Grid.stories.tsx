import React, { ReactElement } from 'react'
import styled from 'styled-components'

import { Grid } from 'components/atoms/Grid'
import { Card } from 'components/atoms/Card'

const StyledCard = styled(Card)`
  display: flex;
  justify-content: center;
`

export const usage = (): ReactElement => (
  <>
    <Grid gap={12}>
      <Grid.Column>
        <Card>1</Card>
      </Grid.Column>
      <Grid.Column>
        <Card>2</Card>
      </Grid.Column>
      <Grid.Column>
        <Card>3</Card>
      </Grid.Column>
      <Grid.Column>
        <Card>4</Card>
      </Grid.Column>
      <Grid.Column>
        <Card>5</Card>
      </Grid.Column>
      <Grid.Column>
        <Card>6</Card>
      </Grid.Column>
      <Grid.Column>
        <Card>7</Card>
      </Grid.Column>
      <Grid.Column>
        <Card>8</Card>
      </Grid.Column>
      <Grid.Column>
        <Card>9</Card>
      </Grid.Column>
      <Grid.Column>
        <Card>10</Card>
      </Grid.Column>
      <Grid.Column>
        <Card>11</Card>
      </Grid.Column>
      <Grid.Column>
        <Card>12</Card>
      </Grid.Column>
    </Grid>
  </>
)

export const withBreakPoints = (): ReactElement => (
  <Grid gap={12}>
    <Grid.Column sm={12} md={6}>
      <StyledCard>sm=12 md=6</StyledCard>
    </Grid.Column>
    <Grid.Column sm={12} md={6}>
      <StyledCard>sm=12 md=6</StyledCard>
    </Grid.Column>

    <Grid.Column sm={12} md={4}>
      <StyledCard>sm=12 md=4</StyledCard>
    </Grid.Column>

    <Grid.Column sm={12} md={4}>
      <StyledCard>sm=12 md=4</StyledCard>
    </Grid.Column>

    <Grid.Column sm={12} md={4}>
      <StyledCard>sm=12 md=4</StyledCard>
    </Grid.Column>
  </Grid>
)
