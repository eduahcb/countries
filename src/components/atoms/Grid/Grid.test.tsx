import React from 'react'
import { render } from '@testing-library/react'

import { Grid } from './Grid'

describe('Grid', () => {
  test('should match snapshot', () => {
    const { asFragment } = render(
      <Grid>
        <div>test</div>
      </Grid>
    )

    expect(asFragment()).toMatchSnapshot()
  })

  test('should match snapshot with gap prop', () => {
    const { asFragment } = render(
      <Grid gap={12}>
        <div>test</div>
      </Grid>
    )

    expect(asFragment()).toMatchSnapshot()
  })

  test.each([
    ['sm', 12],
    ['md', 6],
    ['lg', 4],
    ['xl', 3]
  ])('should match snapshot with %s', (size, value) => {
    const props = {
      [size]: value
    }

    const { asFragment } = render(
      <Grid>
        <Grid.Column {...props}>test</Grid.Column>
      </Grid>
    )

    expect(asFragment()).toMatchSnapshot()
  })
})
