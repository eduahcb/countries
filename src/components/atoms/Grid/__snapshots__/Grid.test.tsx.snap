// Vitest Snapshot v1

exports[`Grid > should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-gswNZR gMdRZi"
  >
    <div>
      test
    </div>
  </div>
</DocumentFragment>
`;

exports[`Grid > should match snapshot with gap prop 1`] = `
<DocumentFragment>
  <div
    class="sc-gswNZR bigLVm"
  >
    <div>
      test
    </div>
  </div>
</DocumentFragment>
`;

exports[`Grid > should match snapshot with lg 1`] = `
<DocumentFragment>
  <div
    class="sc-gswNZR gMdRZi"
  >
    <div
      class="sc-bcXHqe dYjYAm"
    >
      test
    </div>
  </div>
</DocumentFragment>
`;

exports[`Grid > should match snapshot with md 1`] = `
<DocumentFragment>
  <div
    class="sc-gswNZR gMdRZi"
  >
    <div
      class="sc-bcXHqe jVfKEF"
    >
      test
    </div>
  </div>
</DocumentFragment>
`;

exports[`Grid > should match snapshot with sm 1`] = `
<DocumentFragment>
  <div
    class="sc-gswNZR gMdRZi"
  >
    <div
      class="sc-bcXHqe eCBGWj"
    >
      test
    </div>
  </div>
</DocumentFragment>
`;

exports[`Grid > should match snapshot with xl 1`] = `
<DocumentFragment>
  <div
    class="sc-gswNZR gMdRZi"
  >
    <div
      class="sc-bcXHqe edoFpQ"
    >
      test
    </div>
  </div>
</DocumentFragment>
`;
